### ISSUE TITLE
Implement a DrainSchedd

### ISSUE THREAD

**Issue by amaltaro** [dmwm/WMCore#8414]
The difference w.r.t. the standard draining is that DrainSchedd would set all pending thresholds to 0, such that we get jobs out of the global pool and no jobs are submitted to it (thus keeping it all under the internal wmagent database).

Of course the agent cannot pull any work from global or local workqueue.

**Comment by vlimant** [comment]
while you are at it, I would add that the agent could condor_rm all pendings, to get them back warm in local. That way you have better chance that there are no running jobs when you need to pull the plug on the schedd, and none of the pending jobs will end up orphan.
condor_rm should be done in stage to not shoot the schedd either.

**Comment by amaltaro** [comment]
Yes, removing pending jobs a couple of hours before the outage would be a plus, though it might not be always wanted, so it would be better to make that optional. 
For now, I'd rather leave the condor_rm out of the game (or one can do it manually...) because it'd complicate this initial idea and take an additional substantial time.

**Comment by bbockelm** [comment]
Why would pending jobs end up orphans?  Everything should be recovered cleanly after reboots -- things will only get lost in the case of something like filesystem issues.

One thing worth doing would be adding something like:

```
Requirements = ( ... existing stuff ... ) && (HostDraining =?= true)
```

and set `HostDraining` to true for all pending jobs when you want the drain to start (and vice-versa when you come out of draining).  The job edit can be done in a single transaction, so would be far more lightweight than a `condor_rm` (which can trigger some extra IO in the cleanup).

**Comment by vlimant** [comment]
would pending jobs in htcondor survive a schedd full reboot ?

**Comment by bbockelm** [comment]
Ah - I forgot, WMAgent deploys (at least at CERN) move half of the HTCondor state to ramdisk.  Awhile back, this would have been useful to avoid expensive FS operations - can't imagine it being all that useful now.

Since we partially lobotomize HTCondor on restart, I guess the agent should assume all state is lost on reboot.

It's a shame, really, as HTCondor should be able to do simple reboots and reconnect to the running jobs (i.e., the HTCondor side doesn't need to be drained for short outages).

**Comment by amaltaro** [comment]
Instead of condor_rm, I think we could bump all those jobs priority by 100, making sure that jobs from that schedd will go ahead the others. Do you think this would be somehow heavy to condor, a code like
```
fetch all WMAgent_SubTaskName and JobPrio
for each SubTaskName
  condor edit jobs with JobPrio+=100
```

**Comment by vlimant** [comment]
no, it is not realistic to drain the schedd by having all pending running to completion. it could take days, if not week.
@bbockelm do I understand that the HostDraining does not function because of the htcondor reboot on the machine ?

**Comment by amaltaro** [comment]
We're probably not going to save/complete all jobs, but the impact will be much smaller than condor_rm'ing jobs or not touching anything related to the job/schedd priority.

Checking historical dashboard, we seem to be completing ~200k jobs per day, while we usually have between 200k and 300k production jobs pending. Besides the low priority workflows, 2 to 3 days should suffice to get most of them done.

**Comment by bbockelm** [comment]
HostDraining should still work - you just lose all the jobs on reboot and WMAgent would have to resubmit.  That part is just really ineffficient, but would keep new jobs from starting.

**Comment by bbockelm** [comment]
@amaltaro - it's not a standard feature, it's something I proposed a few comments above.

**Comment by vlimant** [comment]
@bbockelm is there really a huge cost to condor_rm 100 jobs every 10s or so ? I'd rather have the agent handle the disconnected jobs gracefully : i.e. condor_rm and flag with a dedicated error code, so as to not flowed operation with 10s of thousands of job with a weird htcondor error. i.e. we see and know what happened to these jobs and put it aside (even better, do not show them as cooloff)

**Comment by amaltaro** [comment]
Going through this thread again, it seems we have deviated from the original issue.

The `DrainSchedd` flag is supposed to have the following impact:
* not submit ANY new jobs to the condor pool (thus draining the pool)
* not acquire ANY new workqueue elements from GQ

The problem is that this schedd draining might take days/weeks, depending on the JobPrio of these jobs and jobs in the pool. Thus we could also bump the JobPrio of all pending jobs on that specific schedd. This draining would probably be very uncommon, so I don't think we should worry about running low prio work before higher ones.

If we think this development won't be that useful, I'm happy to simply close this issue as well. @bbockelm @vlimant
